swim/join_delayer.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package swim
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/uber-common/bark"
+)
+
+var errAttemptNumberInvalid = errors.New("joinDelayer requires attemptNumber to be greater or equal to 0")
+var errLoggerRequired = errors.New("joinDelayer requires a logger")
+
+const (
+	defaultInitial = float64(100 * time.Millisecond)
+	defaultMax     = float64(60 * time.Second)
+)
+
+var defaultRandomizer = rand.Intn
```

Comment:
this randomizer uses the default source that has deterministic behavior. Lets make our own [`Rand`](https://golang.org/pkg/math/rand/#New) with a newly seeded [`Source`](https://golang.org/pkg/math/rand/#NewSource). To prevent interference with the `Source` of the application we are embedded in -that application might rely on the fact that random numbers are deterministic- and at the same time prevent all ringpops to use the same 'random' timeouts.